File: src/main/java/io/smallrye/graphql/clientplugin/Schema.java
package io.smallrye.graphql.clientplugin;

import com.intellij.ide.impl.ProjectUtil;
import com.intellij.lang.java.lexer.JavaLexer;
import com.intellij.pom.java.LanguageLevel;
import graphql.language.DescribedNode;
import graphql.language.FieldDefinition;
import graphql.language.InputValueDefinition;
import graphql.language.ListType;
import graphql.language.NamedNode;
import graphql.language.Node;
import graphql.language.NonNullType;
import graphql.language.SDLNamedDefinition;
import graphql.language.Type;
import graphql.language.TypeDefinition;
import graphql.schema.idl.SchemaParser;
import graphql.schema.idl.TypeDefinitionRegistry;
import graphql.schema.idl.errors.SchemaProblem;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.function.Consumer;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toSet;

class Schema {
    private static final Logger LOG = Logger.getLogger(Schema.class.getName());

    static final Schema INSTANCE = new Schema();

    private TypeDefinitionRegistry cache;
    private Instant lastChange;
    private Consumer<String> errors;

    public Schema withErrors(Consumer<String> errors) {
        this.errors = errors;
        return this;
    }

    public Set<String> typeNames() {
        return schema()
            .stream()
            .flatMap(schema -> schema.types().values().stream())
            .map(SDLNamedDefinition::getName)
            .collect(toSet());
    }

    public Stream<DefinedFieldOrInputValue> fieldsIn(String typeName) {
        return schema()
            .flatMap(schema -> schema.getType(typeName))
            .stream()
            .flatMap(DefinedFieldOrInputValue::in);
    }

    public Optional<TypeDefinitionRegistry> schema() {
        var activeProject = ProjectUtil.getActiveProject();
        if (activeProject == null) return emptyBecause("no active project");
        var basePath = activeProject.getBasePath();
        if (basePath == null) return emptyBecause("no base path in project " + activeProject.getName());
        var path = Paths.get(basePath).resolve("schema.graphql");
        if (!Files.exists(path)) return emptyBecause("no GraphQL schema found at " + path);

        try {
            return Optional.of(parse(path));
        } catch (SchemaProblem e) {
            return emptyBecause("parsing of schema at " + path + " failed:\n" + e.getMessage());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private Optional<TypeDefinitionRegistry> emptyBecause(String message) {
        debug(message);
        errors.accept(message);
        return Optional.empty();
    }

    private TypeDefinitionRegistry parse(Path path) throws IOException {
        var modified = Files.getLastModifiedTime(path).toInstant();
        if (cache == null || modified.isAfter(lastChange)) {
            debug("reload schema");
            cache = new SchemaParser().parse(path.toFile());
            lastChange = modified;
        }
        return cache;
    }

    private static void debug(String message) {
        LOG.fine(message);
    }

    public static class DefinedFieldOrInputValue {
        private static Stream<DefinedFieldOrInputValue> in(TypeDefinition<?> typeDefinition) {
            return typeDefinition.getChildren().stream()
                .filter(node -> node instanceof FieldDefinition || node instanceof InputValueDefinition)
                .map(node -> new DefinedFieldOrInputValue(typeDefinition.getName(), node));
        }

        private final String containerType;
        private final String name;
        private final String description;
        private final Type<?> type;
        private final List<InputValueDefinition> inputValueDefinitions;

        public DefinedFieldOrInputValue(String containerType, Node<?> node) {
            this.containerType = containerType;
            this.name = ((NamedNode<?>) node).getName();
            var description = ((DescribedNode<?>) node).getDescription();
            this.description = (description == null) ? "" : description.getContent();
            if (node instanceof FieldDefinition) {
                var field = (FieldDefinition) node;
                this.type = field.getType();
                this.inputValueDefinitions = field.getInputValueDefinitions();
            } else {
                var input = (InputValueDefinition) node;
                this.type = input.getType();
                this.inputValueDefinitions = List.of();
            }
        }

        public String getQueryType() {
            return containerType;
        }

        public String getName() {
            return name;
        }

        public String getDescription() {
            return (description == null) ? "" : description;
        }

        public String javaMethod() {
            return "@" + getQueryType() + " " + toJava(type) + " " + getName() + "(" + parameters() + ");";
        }

        public String javaField() {
            return toJava(type) + " " + getName() + ";";
        }

        private static String toJava(Type<?> type) {
            var prefix = "";
            var suffix = "";
            if (type instanceof NonNullType) {
                prefix += "@NonNull ";
                type = ((NonNullType) type).getType();
            }
            if (type instanceof ListType) {
                prefix += "List<";
                suffix += ">";
                type = ((ListType) type).getType();
            }
            if (type instanceof NonNullType) {
                prefix += "@NonNull ";
                type = ((NonNullType) type).getType();
            }
            var typeName = ((NamedNode<?>) type).getName();
            switch (typeName) {
                case "Int":
                    typeName = "Integer";
                    break;
                case "ID":
                    typeName = "String";
                    prefix = "@Id " + prefix;
                    break;
            }
            return prefix + typeName + suffix;
        }

        private String parameters() {
            return inputValueDefinitions.stream()
                .map(DefinedFieldOrInputValue::parameterDeclaration)
                .collect(Collectors.joining(", "));
        }

        private static String parameterDeclaration(InputValueDefinition param) {
            var annotations = "";
            var paramName = param.getName();
            if (JavaLexer.isKeyword(paramName, LanguageLevel.HIGHEST)) {
                annotations = "@Name(\"" + paramName + "\") ";
                paramName = paramName + "_";
            }
            return annotations + toJava(param.getType()) + " " + paramName;
        }
    }
}

File: src/main/kotlin/io/smallrye/graphql/clientplugin/Schema.kt
package io.smallrye.graphql.clientplugin

import com.intellij.ide.impl.ProjectUtil
import com.intellij.lang.java.lexer.JavaLexer
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.NotificationType.ERROR
import com.intellij.notification.NotificationType.WARNING
import com.intellij.notification.Notifications
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.pom.java.LanguageLevel
import graphql.language.DescribedNode
import graphql.language.FieldDefinition
import graphql.language.InputValueDefinition
import graphql.language.ListType
import graphql.language.NamedNode
import graphql.language.Node
import graphql.language.NonNullType
import graphql.language.Type
import graphql.language.TypeDefinition
import graphql.schema.idl.SchemaParser
import graphql.schema.idl.TypeDefinitionRegistry
import io.ktor.util.collections.ConcurrentSet
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.time.Instant
import java.util.logging.Logger
import java.util.stream.Collectors
import java.util.stream.Stream

internal class Schema {
    private var cache: TypeDefinitionRegistry? = null
    private var lastChange: Instant? = null

    fun typeNames(): Set<String> {
        return schema()
            ?.types()
            ?.values
            ?.map { obj -> obj.name }
            ?.toSet()
            ?: emptySet()
    }

    fun fieldsIn(typeName: String): Stream<DefinedFieldOrInputValue> {
        return schema()
            ?.getType(typeName)
            ?.stream()
            ?.flatMap { typeDefinition -> DefinedFieldOrInputValue.of(typeDefinition) }
            ?: Stream.empty()
    }

    private fun schema(): TypeDefinitionRegistry? {
        val activeProject = ProjectUtil.getActiveProject() ?: return noActiveProject()
        val basePath = activeProject.basePath ?: return noBasePathInProject(activeProject)
        val path = Paths.get(basePath).resolve("schema.graphql")
        return if (!Files.exists(path)) noGraphQLSchemaFound(path)
        else try {
            parse(path)
        } catch (e: RuntimeException) {
            schemaParsingFailed(path, e)
        }
    }

    private fun noActiveProject(): TypeDefinitionRegistry? {
        notify(ERROR, "no active project")
        return null
    }

    private fun noBasePathInProject(activeProject: Project): TypeDefinitionRegistry? {
        notify(ERROR, "no base path in project " + activeProject.name)
        return null
    }

    private fun schemaParsingFailed(path: Path?, e: RuntimeException): TypeDefinitionRegistry? {
        notify(ERROR, "parsing of schema at $path failed:\n${e.message}")
        return null
    }

    private fun parse(path: Path): TypeDefinitionRegistry {
        val modified = Files.getLastModifiedTime(path).toInstant()
        if (cache == null || modified.isAfter(lastChange)) {
            debug("reload schema")
            cache = SchemaParser().parse(path.toFile())
            lastChange = modified
            SHOWN_NOTIFICATIONS.clear()
        }
        return cache!!
    }

    private fun noGraphQLSchemaFound(path: Path): TypeDefinitionRegistry? {
        notify(WARNING, "no GraphQL schema found at $path")
        return null
    }

    private fun notify(notificationType: NotificationType, message: String) {
        debug(message)
        if (SHOWN_NOTIFICATIONS.add(message)) {
            Notifications.Bus.notify(Notification("GraphQL Client Notifications", message, notificationType))
        }
    }

    internal class DefinedFieldOrInputValue(val queryType: String, node: Node<*>) {
        val name: String
        val description: String
        private var type: Type<*>
        private var inputValueDefinitions: List<InputValueDefinition>

        init {
            name = (node as NamedNode<*>).name
            val description = (node as DescribedNode<*>).description
            this.description = if (description == null) "" else description.getContent()
            if (node is FieldDefinition) {
                type = node.type
                inputValueDefinitions = node.inputValueDefinitions
            } else {
                val input = node as InputValueDefinition
                type = input.type
                inputValueDefinitions = listOf()
            }
        }

        fun javaMethod(): String {
            return "@" + queryType + " " + toJava(type) + " " + name + "(" + parameters() + ");"
        }

        fun javaField(): String {
            return toJava(type) + " " + name + ";"
        }

        private fun parameters(): String {
            return inputValueDefinitions.stream()
                .map { param -> parameterDeclaration(param) }
                .collect(Collectors.joining(", "))
        }

        companion object {
            internal fun of(typeDefinition: TypeDefinition<*>): Stream<DefinedFieldOrInputValue> {
                return typeDefinition.children.stream()
                    .filter { node -> node is FieldDefinition || node is InputValueDefinition }
                    .map { node -> DefinedFieldOrInputValue(typeDefinition.name, node) }
            }

            private fun toJava(type: Type<*>): String {
                var result = type
                var prefix = ""
                var suffix = ""
                if (result is NonNullType) {
                    prefix += "@NonNull "
                    result = result.type
                }
                if (result is ListType) {
                    prefix += "List<"
                    suffix += ">"
                    result = result.type
                }
                if (result is NonNullType) {
                    prefix += "@NonNull "
                    result = result.type
                }
                var typeName = (result as NamedNode<*>).name
                when (typeName) {
                    "Int" -> typeName = "Integer"
                    "ID" -> {
                        typeName = "String"
                        prefix = "@Id $prefix"
                    }
                }
                return prefix + typeName + suffix
            }

            private fun parameterDeclaration(param: InputValueDefinition): String {
                var annotations = ""
                var paramName = param.name
                if (JavaLexer.isKeyword(paramName, LanguageLevel.HIGHEST)) {
                    annotations = "@Name(\"$paramName\") "
                    paramName += "_"
                }
                return annotations + toJava(param.type) + " " + paramName
            }
        }
    }

    companion object {
        lateinit var editor: Editor
        val INSTANCE: Schema = Schema()
        private val SHOWN_NOTIFICATIONS = ConcurrentSet<String>()
        private val LOG = Logger.getLogger(Schema::class.java.name)
        private fun debug(message: String) {
            LOG.fine(message)
        }
    }
}
